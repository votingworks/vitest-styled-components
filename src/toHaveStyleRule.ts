import { CssAtRuleAST, CssDeclarationAST, CssStylesheetAST, CssTypes } from '@adobe/css-tools';
import type * as sc from 'styled-components';
import type { ExpectStatic } from 'vitest';
import { areCssMediaQueriesEqual, areCssSupportsEqual, buildReturnMessage, getCSS, matcherTest } from './utils.js';

export type Value = string | RegExp | boolean | null | undefined;

export interface Options {
  media?: string;
  modifier?: string | sc.RuleSet;
  supports?: string;
}

interface NormalizedOptions {
  media?: string;
  modifier?: string;
  supports?: string;
}

export type ToHaveStyleRuleMatcher = (
  element: Element,
  property: string,
  expected?: Value,
  options?: Options,
) => { pass: boolean; message: () => string };

const isStyledClass = (className: string): boolean => /(_|-)+sc-.+|^sc-/.test(className);

const getClassNames = (received?: Element): string[] => {
  return received?.classList ? Array.from(received.classList) : [];
};

function hasAtRule(options: Options): boolean {
  return Boolean(options.media || options.supports);
}

function getAtRules(ast: CssStylesheetAST, options: Options) {
  let rules = ast.stylesheet.rules;
  const { media, supports } = options;

  if (media) {
    rules = rules.filter((rule) => rule.type === CssTypes.media && areCssMediaQueriesEqual(rule.media, media));
  }

  if (supports) {
    rules = rules.filter((rule) => rule.type === CssTypes.supports && areCssSupportsEqual(rule.supports, supports));
  }

  return rules;
}

/** stylis v4 renders descendant selectors without a trailing space sometimes which trips up detection */
function removeSpaceAfterSelector(input: string): string {
  return input.replace(/([>~+]) +/g, '$1');
}

function normalizeQuotations(input: string): string {
  return input.replace(/['"]/g, '"');
}

function getModifiedClassName(className: string, staticClassName: string, modifier = ''): string {
  const classNameSelector = `.${className}`;
  let prefix = '';

  modifier = modifier.trim();
  if (modifier.includes('&')) {
    modifier = modifier
      // & combined with other selectors and not a precedence boost should be replaced with the static className, but the first instance should be the dynamic className
      .replace(/(&[^&]+?)&/g, `$1.${staticClassName}`)
      .replace(/&/g, classNameSelector);
  } else {
    prefix += classNameSelector;
  }
  const first = modifier[0];
  if (first !== ':' && first !== '[') {
    prefix += ' ';
  }

  return `${prefix}${modifier}`.trim();
}

function hasClassNames(
  classNames: readonly string[],
  selectors: readonly string[],
  options: NormalizedOptions,
): boolean {
  const staticClassNames = classNames.filter((x) => isStyledClass(x));

  return classNames.some((className) =>
    staticClassNames.some((staticClassName) =>
      selectors.map(removeSpaceAfterSelector).includes(
        removeSpaceAfterSelector(
          normalizeQuotations(
            getModifiedClassName(className, staticClassName, options.modifier).replace(/['"]/g, '"')
          )
        )
      )
    )
  );
}

function getRules(ast: CssStylesheetAST, classNames: readonly string[], options: NormalizedOptions) {
  if (hasAtRule(options)) {
    return getAtRules(ast, options);
  }

  return ast.stylesheet.rules
    .filter((rule) => rule.type === CssTypes.rule)
    .filter((rule) =>
      'selectors' in rule && hasClassNames(classNames, rule.selectors.map(normalizeQuotations), options)
    );
}

function handleMissingRules(options: Options) {
  return ({
    pass: false,
    message: () =>
      `No style rules found on passed element${
        Object.keys(options).length ? ` using options:\n${JSON.stringify(options)}` : ''
      }`,
  });
}

function getDeclaration(rule: CssAtRuleAST, property: string): CssDeclarationAST | undefined {
  switch (rule.type) {
    case CssTypes.rule:
      return rule.declarations
        .filter((declaration): declaration is CssDeclarationAST => declaration.type === CssTypes.declaration)
        .filter((declaration) => declaration.property === property)
        .pop();

    case CssTypes.supports:
    case CssTypes.media: {
      for (const subrule of rule.rules) {
        const declaration = getDeclaration(subrule, property);
        if (declaration) {
          return declaration;
        }
      }
      break;
    }
  }
}

function getDeclarations(rules: readonly CssAtRuleAST[], property: string) {
  return rules.reduce<CssDeclarationAST[]>((acc, rule) => {
    const declaration = getDeclaration(rule, property);
    return declaration ? [...acc, declaration] : acc;
  }, []);
}

function normalizeOptions(options: Options): NormalizedOptions {
  return {
    ...options,
    modifier: Array.isArray(options.modifier) ? options.modifier.join('') : options.modifier,
  };
}

export interface ToHaveStyleRuleMatchers {
  toHaveStyleRule: Parameters<ToHaveStyleRuleMatcher> extends [unknown, ...infer R] ? (...args: R) => void : never;
}

/**
 * Builds the `toHaveStyleRule` matcher given an `expect` function.
 */
export function buildToHaveStyleRule(expect: ExpectStatic): ToHaveStyleRuleMatcher {
  return function(this: { isNot: boolean }, element, property, expected, options = {}) {
    const classNames = getClassNames(element);
    const ast = getCSS();
    const normalizedOptions = normalizeOptions(options);
    const rules = getRules(ast, classNames, normalizedOptions);

    if (!rules.length) {
      return handleMissingRules(normalizedOptions);
    }

    const declarations = getDeclarations(rules, property);
    const received = declarations.pop()?.value;
    const pass = matcherTest(expect, received, expected, this.isNot);

    return {
      pass,
      message: buildReturnMessage(pass, property, received, expected),
    };
  };
}
