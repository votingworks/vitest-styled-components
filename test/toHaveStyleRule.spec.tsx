import { render } from '@testing-library/react';
import React from 'react';
import { css, styled, ThemeProvider } from 'styled-components';
import { expect, test } from 'vitest';
import { Options, Value } from '../src/toHaveStyleRule.js';

function rendered(component: React.ReactNode) {
  return render(component).container.firstChild;
}

function assertHasStyleRule(component: React.ReactNode, property: string, value: Value, options?: Options) {
  expect(rendered(component)).toHaveStyleRule(property, value, options);
}

function assertDoesNotHaveStyleRule(component: React.ReactNode, property: string, value?: Value) {
  expect(rendered(component)).not.toHaveStyleRule(property, value);
}

test('null', () => {
  expect(null).not.toHaveStyleRule('a', 'b');
});

test('non-styled', () => {
  assertDoesNotHaveStyleRule(<div />, 'a', 'b');
});

test('message when rules not found', () => {
  expect(() => assertHasStyleRule(<div />, 'color', 'black'))
    .toThrowError('No style rules found on passed element');
});

test('message when rules not found using options', () => {
  const Button = styled.button`
    color: red;
  `;

  assertHasStyleRule(<Button />, 'color', 'red');
  expect(() =>
    assertHasStyleRule(<Button />, 'color', 'red', {
      media: '(max-width:640px)',
      modifier: ':hover',
    })
  ).toThrowError('No style rules found on passed element using options');
});

test('message when property not found', () => {
  const Button = styled.button`
    color: red;
  `;

  expect(() => assertHasStyleRule(<Button />, 'background-color', 'black'))
    .toThrowError(`Property 'background-color' not found in style rules`);
});

test('message when value does not match', () => {
  const Wrapper = styled.section`
    background: orange;
  `;

  expect(() => {
    assertHasStyleRule(<Wrapper />, 'background', 'red');
  }).toThrowError(`Value mismatch for property 'background'`);
});

test('non existing', () => {
  expect(() => {
    expect(render(<div />)).toHaveStyleRule('background', 'papayawhip');
  }).toThrowError('No style rules found on passed element');
});

test('basic', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `;

  assertHasStyleRule(<Wrapper />, 'background', 'papayawhip');
});

test('regex', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `;

  assertHasStyleRule(<Wrapper />, 'background', /^p/);
});

test('complex string', () => {
  const Wrapper = styled.section`
    border: 1px solid rgba(0, 0, 0, 0.125);
  `;

  assertHasStyleRule(<Wrapper />, 'border', '1px solid rgba(0,0,0,0.125)');
});

test('undefined', () => {
  const Button = styled.button`
    cursor: ${({ disabled }) => !disabled && 'pointer'};
    opacity: ${({ disabled }) => disabled && '.65'};
  `;

  assertHasStyleRule(<Button />, 'opacity', undefined);
  assertHasStyleRule(<Button />, 'cursor', 'pointer');
  assertHasStyleRule(<Button disabled />, 'opacity', '.65');
  assertHasStyleRule(<Button disabled />, 'cursor', undefined);
});

test('negated ".not" modifier with no value', () => {
  const Button = styled.button`
    opacity: ${({ disabled }) => disabled && '.65'};
  `;

  assertDoesNotHaveStyleRule(<Button />, 'opacity');
  assertHasStyleRule(<Button disabled />, 'opacity', '.65');
});

test('negated ".not" modifier with value', () => {
  const Button = styled.button`
    opacity: 0.65;
  `;

  assertDoesNotHaveStyleRule(<Button />, 'opacity', '0.50');
  assertDoesNotHaveStyleRule(<Button />, 'opacity', '');
  assertDoesNotHaveStyleRule(<Button />, 'opacity', null);
  assertDoesNotHaveStyleRule(<Button />, 'opacity', false);
  expect(() => {
    assertHasStyleRule(<Button />, 'opacity', undefined);
  }).toThrowError();
});

test('negated ".not" modifier fails when rule present with no value being asserted', () => {
  const Button = styled.button`
    opacity: 0.65;
  `;

  assertHasStyleRule(<Button />, 'opacity', '0.65');
  expect(() => {
    assertDoesNotHaveStyleRule(<Button />, 'opacity');
  }).toThrowError();
});

test('asymmetric matchers', () => {
  const Button = styled.button<{ $transparent?: boolean }>`
    border: 0.1em solid ${({ $transparent }) => ($transparent ? 'transparent' : 'black')};
  `;

  assertHasStyleRule(<Button />, 'border', expect.any(String));
  assertHasStyleRule(<Button />, 'border', expect.stringMatching('solid'));
  assertHasStyleRule(<Button />, 'border', expect.stringMatching(/^0.1em/));
  assertHasStyleRule(<Button />, 'border', expect.stringContaining('black'));
  assertDoesNotHaveStyleRule(<Button $transparent />, 'border', expect.stringContaining('black'));
  assertHasStyleRule(<Button $transparent />, 'border', expect.stringContaining('transparent'));
  assertDoesNotHaveStyleRule(<Button />, 'color', expect.any(String));
  assertDoesNotHaveStyleRule(<Button />, 'color', expect.anything());
});

test('any component', () => {
  const Link = ({ className, children }: {
    className?: string;
    children?: React.ReactNode;
  }) => <a className={className}>{children}</a>;

  const StyledLink = styled(Link)`
    color: palevioletred;
    font-weight: bold;
  `;

  assertHasStyleRule(<StyledLink>Styled, exciting Link</StyledLink>, 'color', 'palevioletred');
});

test('styled child', () => {
  const Parent = styled.div`
    color: red;
  `;

  const StyledChild = styled(Parent)`
    padding: 0;
  `;

  assertHasStyleRule(<StyledChild />, 'color', 'red');
});

test('theming', () => {
  const Button = styled.button<{ theme: { main: string } }>`
    font-size: 1em;
    margin: 1em;
    padding: 0.25em 1em;
    border-radius: 3px;

    color: ${(props) => props.theme.main};
    border: 2px solid ${(props) => props.theme.main};
  `;

  Button.defaultProps = {
    theme: {
      main: 'palevioletred',
    },
  };

  const theme = {
    main: 'mediumseagreen',
  };

  assertHasStyleRule(<Button>Normal</Button>, 'color', 'palevioletred');

  const component = (
    <ThemeProvider theme={theme}>
      <Button>Themed</Button>
    </ThemeProvider>
  );

  assertHasStyleRule(component, 'color', 'mediumseagreen');
});

test('at rules', () => {
  const Wrapper = styled.section`
    color: red;
    @media (max-width: 640px) {
      color: green;
    }
    @media (min-width: 200px) and (max-width: 640px) {
      color: blue;
    }
    @media (min-width: 576px) and (max-width: 767.98px) {
      color: red;
    }
    @media (min-width: calc(768px + 1px)) and (max-width:calc(1024px + 1px)) {
      color: purple;
    }
  `;

  assertHasStyleRule(<Wrapper />, 'color', 'red');
  assertHasStyleRule(<Wrapper />, 'color', 'green', {
    media: '(max-width:640px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'green', {
    media: '(max-width: 640px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width:200px) and (max-width:640px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width: 200px) and (max-width: 640px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width: 200px) and (max-width:640px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width:200px) and (max-width: 640px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'red', {
    media: '(min-width: 576px) and (max-width: 767.98px)',
  });
  assertHasStyleRule(<Wrapper />, 'color', 'purple', {
    media: '(min-width: calc(768px + 1px)) and (max-width:calc(1024px + 1px))',
  });
});

test('selector modifiers', () => {
  const Link = styled.a`
    color: white;

    &:hover {
      color: blue;
    }
    &::after {
      color: red;
    }
    &[href*='somelink.com'] {
      color: green;
    }
    > div {
      color: yellow;
    }
    span {
      color: purple;
    }
    .child {
      color: orange;
    }
    &.self {
      color: black;
    }

    .one,
    .two {
      color: olive;
    }

    ~ div {
      &.one,
      &.two {
        color: pink;
      }
    }

    + div {
      .one,
      .two {
        color: salmon;
      }
    }

    .parent & {
      color: red;
    }

    && {
      color: fuchsia;
    }

    &&& {
      color: olive;
    }

    & & {
      color: deepskyblue;
    }
  `;

  assertHasStyleRule(<Link />, 'color', 'white');
  assertHasStyleRule(<Link />, 'color', 'blue', { modifier: ':hover' });
  assertHasStyleRule(<Link />, 'color', 'red', { modifier: '::after' });
  assertHasStyleRule(<Link />, 'color', 'green', { modifier: "[href*='somelink.com']" });
  assertHasStyleRule(<Link />, 'color', 'yellow', { modifier: '> div' });
  assertHasStyleRule(<Link />, 'color', 'purple', { modifier: 'span' });
  assertHasStyleRule(<Link />, 'color', 'purple', { modifier: ' span' });
  assertHasStyleRule(<Link />, 'color', 'orange', { modifier: '.child' });
  assertHasStyleRule(<Link />, 'color', 'orange', { modifier: ' .child' });
  assertHasStyleRule(<Link />, 'color', 'black', { modifier: '&.self' });
  assertHasStyleRule(<Link />, 'color', 'olive', { modifier: '.one' });
  assertHasStyleRule(<Link />, 'color', 'olive', { modifier: '.two' });
  assertHasStyleRule(<Link />, 'color', 'pink', { modifier: '~ div.one' });
  assertHasStyleRule(<Link />, 'color', 'salmon', { modifier: '+ div .two' });
  assertHasStyleRule(<Link />, 'color', 'red', { modifier: '.parent &' });
  assertHasStyleRule(<Link />, 'color', 'fuchsia', { modifier: '&&' });
  assertHasStyleRule(<Link />, 'color', 'olive', { modifier: '&&&' });
  assertHasStyleRule(<Link />, 'color', 'deepskyblue', { modifier: '& &' });
});

test('component modifiers', () => {
  const Text = styled.span`
    color: grey;
  `;

  const Link = styled.a`
    color: white;

    ${Text} {
      color: blue;
    }

    > ${Text} span {
      color: green;
    }

    ${Text} & {
      color: purple;
    }
  `;

  assertHasStyleRule(<Link />, 'color', 'white');
  assertHasStyleRule(<Text />, 'color', 'grey');
  assertHasStyleRule(
    <Link>
      <Text />
    </Link>,
    'color',
    'blue',
    { modifier: css`${Text}` }
  );
  assertHasStyleRule(
    <Link>
      <Text />
    </Link>,
    'color',
    'green',
    { modifier: css`> ${Text} span` }
  );
  assertHasStyleRule(
    <Link>
      <Text />
    </Link>,
    'color',
    'purple',
    { modifier: css`${Text} &` }
  );
});

test('nested', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `;

  const MyComponent = () => <Wrapper />;

  assertHasStyleRule(<MyComponent />, 'background', 'papayawhip');
});

test('nested with styling', () => {
  const Wrapper = styled.section`
    background: papayawhip;
  `;
  const Children = styled.span`
    background: gray;
  `;
  const MyComponent = ({ className, children }: { className?: string; children?: React.ReactNode }) => (
    <Wrapper className={className}>{children}</Wrapper>
  );
  const MyStyledComponent = styled(MyComponent)`
    color: red;
  `;
  const ParentComponent = ({ className }: { className?: string }) => (
    <MyStyledComponent className={className}>
      <Children className='test-class' />
    </MyStyledComponent>
  );

  assertHasStyleRule(<MyStyledComponent />, 'color', 'red');
  assertHasStyleRule(<MyStyledComponent className='test-class' />, 'color', 'red');
  expect(render(<ParentComponent />).container.firstChild?.firstChild).toHaveStyleRule('background', 'gray');
});

test('empty children', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `;

  Wrapper.defaultProps = {
    children: '',
  };

  assertHasStyleRule(<Wrapper />, 'background', 'papayawhip');
});

test('custom display name prefix', () => {
  const Text = styled.span.withConfig({ displayName: 'Text__sc' })`
    color: red;
  `;
  const Comp = styled(Text).withConfig({ displayName: 'Comp__Sub-sc' })`
    background: papayawhip;
  `;
  assertHasStyleRule(<Comp />, 'background', 'papayawhip');
  assertHasStyleRule(<Comp />, 'color', 'red');
});

test('supports snake case display name prefix', () => {
  const Text = styled.span`
    color: blue;
  `;
  Text.styledComponentId = `test-case-${Text.styledComponentId}`;

  assertHasStyleRule(<Text />, 'color', 'blue');
});
