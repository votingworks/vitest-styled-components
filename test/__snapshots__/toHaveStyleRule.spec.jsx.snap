// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`message when property not found 1`] = `[Error: Invalid Chai property: toHaveStyleRule]`;

exports[`message when rules not found 1`] = `[Error: Invalid Chai property: toHaveStyleRule]`;

exports[`message when rules not found using options 1`] = `
[Error: No style rules found on passed Component using options:
{"media":"(max-width:640px)","modifier":":hover"}]
`;

exports[`message when value does not match 1`] = `[Error: Invalid Chai property: toHaveStyleRule]`;

exports[`non existing 1`] = `[Error: Invalid Chai property: toHaveStyleRule]`;
